### ISSUE TITLE
Allow disabling destination

### ISSUE THREAD

**Issue by lijok** [hookdeck/terraform-provider-hookdeck#128]
We need a way to disable destinations via terraform. I see there's a disabled_at option present on `hookdeck_destination` resource, but does not appear to do anything?

**Comment by leggetter** [comment]
With Terraform, you'd expect to be able to set a disabled property on a resource to disable it. However, the Destination, Connection, and Source resources do not have a `disabled` / `enabled` or `paused` property, so instead, we need to trigger a side effect of calling the /disable/enable/pause/unpause RPC method.

@alexluong - what are your thoughts on having a property on the Terraform resource but not on the API resource? Also, could we then hook into the flow, and have a side effect of calling the RPC endpoints?

**Comment by alexluong** [comment]
@lijok hi, any chance you can elaborate on your use case a bit? Currently the TF doesn't support this operation as it's not the intended use case for TF. We can certainly support it but would be curious to hear of your use case if possible. Any reason you prefer managing this with TF over the dashboard?

As for the `disabled_at` property, it is a read-only field so you can't really set it. If we were to support this behavior, we would need to expose a separate field `disabled` since the `disabled_at` shall be the timestamp from the server.

---

@leggetter 

> However, the Destination, Connection, and Source resources do not have a disabled / enabled or paused property

Technically, this is not the case. From the API perspective, there's no `disabled` | `enabled` | `paused` property either. There's the timestamps, and whether it's disabled is derived from whether the timestamp exists.

cc [destination ref](https://hookdeck.com/docs/api#connection-object)

> Also, could we then hook into the flow, and have a side effect of calling the RPC endpoints?

Yes, that's possible but I'm not sure if it's an antipattern or not.

**Comment by alexluong** [comment]
> Now, the issue, though, is that a source and destination can also be disabled/paused, but that has the side effect of also updating the associated connections, which would break the Terraform state.

Ah that's tough. I don't think there's a clear TF best practice because we ventured into a use case where Terraform is not specifically designed for. With that said, this is a common pattern, so I've looked at how other providers approach this.

- [aws_cloudwatch_event_rule](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_event_rule) has a [`state`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/cloudwatch_event_rule#state-2) field for users to disable / enable the rule.
- [google_cloud_scheduler_job](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_scheduler_job) also has a [`paused`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_scheduler_job#example-usage---scheduler-job-paused) boolean property.
- AWS's EC2 instance has a dedicated [`aws_ec2_instance_state`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/ec2_instance_state) resource to manage its state (running or stopped).

Now, none of them has the cascading effect of our source -> connection. However, I think as long as we have clear documentation that users should read the `disabled_at` timestamp to get the correct state, then we should be fine-ish in terms of the cascading issue.

When our source is disabled, upon state refresh at the end of the TF lifecycle, it will correctly update the `disabled_at` state of affected connections.

Personally, I would prefer a dedicated resource to update state because it may cause less confusion to the end users. If there's a `connection.state` or `connection.disabled` property, people may be inclined to read it as the source of truth over `connection.disabled_at`.

## Problem

The main issue with this approach is this flow:

Let's say we have 1 source with 5 connections

1: disable source  -> cascade & disable all 5 connections
2: enable source

End state: source is enabled; all 5 connections are disabled

Depending on our approach, there's either no way to fix the disabled state, or users will have to be explicit about re-enabling them.

Maybe this is not an issue that we need to address (or not yet) while trying to deliver this feature. I've explored a few approaches but nothing cleanly solves this problem just yet.

### Potential solution

Requiring explicit disabled connections when disabling source or destination:

```hcl
resource "hookdeck_source_pause" "pause_source" {
  source_id = hookdeck_source.my_source.id
  connection_pause_ids = [
    hookdeck_connection_pause.pause_connection_1.id,
    hookdeck_connection_pause.pause_connection_2.id
  ]
}
```

The problem with this is we still need to consider whether this array is an exhaustive list of all affected connections or not. And technically, we can query the API and validate, but what if there are connections not managed by Terraform? So we may need to allow non-exhaustive list, but at least it would still encourage a more explicit workflow.

So edge case would be users fail to provide a full list of `hookdeck_connection_pause` resources, and therefore those connections states are stuck afterwards.

Hopefully, this gives a fair idea of some of the potential issues we should consider when it comes to this feature. Let me know if you have any input or what we should consider here.

**Comment by leggetter** [comment]
> Personally, I would prefer a dedicated resource to update state because it may cause less confusion to the end users.

I hear what you're saying. But then you add a level of separation between the resource having its state changed and the action of applying that change. I think that would provide a poorer DX from a Terraform management perspective and that adding the virtual properties directly to the resources is a better DX.

> [google_cloud_scheduler_job](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_scheduler_job) also has a [paused](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_scheduler_job#example-usage---scheduler-job-paused) boolean property.

The Google Cloud Scheduler is also an RPC method, so it is relevant to this scenario:
https://cloud.google.com/scheduler/docs/reference/rest/v1beta1/projects.locations.jobs/pause

Clearly, Google is also working around this with their Terraform provider.

> Let's say we have 1 source with 5 connections
> 
> 1: disable source -> cascade & disable all 5 connections
> 2: enable source
> 
> End state: source is enabled; all 5 connections are disabled

If this is how the API behaves, then that's outside of the Terraform provider. i.e., not something we're trying to solve here.

My suggestion would be:

1. Add `disabled` and `paused` properties to the various resources
2. On apply, catch those properties and map them to the appropriate RPC API calls (assuming some state change logic would also be applied)
3. (not sure if this is possible or relevant in TF): when retrieving the value of `disabled` and `paused`, use the value of the `{action}_at` timestamp from the API to determine the value.
4. Leave the cascading effects of disabling and enabling Connections to the API and read the resulting values after those have taken effect.

**Comment by leggetter** [comment]
> [@leggetter](https://github.com/leggetter) I hear you, let me explore this approach a bit with code and see how things come together!

Good call. It may flag other challenges and identify what is and is not possible.

**Comment by alexluong** [comment]
Question: what about this scenario: in Terraform, the user has source.disabled = true and connected connection.disabled = false _explicitly_?

> when retrieving the value of disabled and paused, use the value of the {action}_at timestamp from the API to determine the value.

You're suggesting this is the behavior of a computed value. I'm not sure if something can be both explicit & computed at the same time.

**Comment by leggetter** [comment]
@alexluong, what would the state across the Source and Connection be if you were to do the following via the API:

1. Disable the Source
2. Enable the Connection

?

As mentioned earlier, we should leave the Source/Destination/Connection disable logic to the API and just translate the results to the resource properties.

I'm not too sure of the TF lifecycle so maybe is causes a problem?

**Comment by alexluong** [comment]
> what would the state across the Source and Connection be if you were to do the following via the API: 1. Disable the Source 2. Enable the Connection?

Then both would become enabled.

> As mentioned earlier, we should leave the Source/Destination/Connection disable logic to the API and just translate the results to the resource properties.

My concern is when someone sets the field _explicitly_, i.e. hard-coding the value in the Terraform code themselves. What if they set an impossible state?

So our idea is if someone selects set the field `disabled = true`, we'll make an API call to make that property disabled in the API. But what if specifically set the source to be disabled but connection enabled at the same time? We cannot make it so that state exists.

**Comment by leggetter** [comment]
Maybe we should only allow setting Connection state within Terrafor? Would this avoid the problem?

If someone changes the state of a Connection, does it cascade down to the Source? Even if that Source is used within other Connections?

**Comment by alexbouchardd** [comment]
> Maybe we should only allow setting Connection state within TF and not Source state? Would this avoid the problem? 

This ^

**Comment by alexluong** [comment]
> Maybe we should only allow setting Connection state within TF and not Source state? Would this avoid the problem?

Yes, this should solve the issue in that case.
